Public/Src/Tools/VBCSCompilerLogger/VBCSCompilerLogger.cs

```diff
@@ -0,0 +1,191 @@
+using System.Diagnostics.ContractsLight;
+using System.IO;
+using System.Linq;
+using BuildXL.Native.IO;
+using BuildXL.Processes;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+using Microsoft.CodeAnalysis;
+
+namespace VBCSCompilerLogger
+{
+    /// <summary>
+    /// This logger catches csc and vbc MSBuild tasks and uses the command line argument passed to the compiler to mimic the file accesses that the compiler 
+    /// would have produced.
+    /// </summary>
+    /// <remarks>
+    /// This MSBuild logger is used by the MSBuild frontend of BuildXL when scheduling pips: when process breakaway is supported, VBSCompiler.exe is allowed to escape the sandbox
+    /// and outlive the originating pip. This logger is used as a way to compensate for the missing accesses and the MSBuild frontend attaches it to every MSBuild invocation.
+    /// </remarks>
+    public class CompilerFileAccessLogger : Logger
+    {
+        private const string CscTaskName = "Csc";
+        private const string VbcTaskName = "Vbc";
+        private const string CscToolName = "csc.exe";
+        private const string VbcToolName = "vbc.exe";
+        private AugmentedManifestReporter m_augmentedReported;
+
+        /// <inheritdoc/>
+        public override void Initialize(IEventSource eventSource)
+        {
+            eventSource.MessageRaised += EventSourceOnMessageRaised;
+            m_augmentedReported = AugmentedManifestReporter.Instance;
+        }
+
+        private void EventSourceOnMessageRaised(object sender, BuildMessageEventArgs e)
+        {
+            if (e is TaskCommandLineEventArgs commandLine)
+            {
+                // We are only interested in CSharp and VisualBasic tasks
+                string language;
+                string arguments;
+                switch (commandLine.TaskName)
+                {
+                    case CscTaskName:
+                        language = LanguageNames.CSharp;
+                        arguments = GetArgumentsFromCommandLine(CscToolName, commandLine.CommandLine);
+                        break;
+                    case VbcTaskName:
+                        language = LanguageNames.VisualBasic;
+                        arguments = GetArgumentsFromCommandLine(VbcToolName, commandLine.CommandLine);
+                        break;
+                    default:
+                        return;
+                }
+
+                // We were able to split the compiler invocation from its arguments. This is the indicator
+                // that something didn't go as expected. Since failing to parse the command line means we
+                // are not catching all inputs/outputs properly, we have no option but to fail the corresponding pip
+                if (arguments == null)
+                {
+                    throw new ArgumentException($"Unexpected tool name in command line. Expected '{CscToolName}' or '{VbcToolName}', but got: {commandLine.CommandLine}");

```

Comment:
nit: method makes assumptions on the root cause of the received null.

Reply:
Fair, I'll refactor that


---
In reply to: [346063932](https://github.com/microsoft/BuildXL/pull/1207#discussion_r346063932) [](ancestors = 346063932)